From b54c2e2158c1b0c137d3e830ed7794c9ad8d7057 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 22 Nov 2012 21:30:59 -0500
Subject: [PATCH] Misc EMC Changes

Only spawn 1 ocelot ever from eggs.
Load custom permissions before plugins
Proper colors in console
Unload plugins after everything else
Fix boat wood gen exploit
Remove CraftBukkit "Feature" of ItemStack's having infinite use if less than 0 stack size.
Color name tags to match list
Use 1.4.5 Dig Speed
---
 src/main/java/net/minecraft/server/EntityBoat.java           |  2 +-
 src/main/java/net/minecraft/server/EntityHuman.java          | 11 ++---------
 src/main/java/net/minecraft/server/EntityOcelot.java         |  2 +-
 src/main/java/net/minecraft/server/EntityTrackerEntry.java   |  5 +++++
 src/main/java/net/minecraft/server/MinecraftServer.java      | 12 +++++++-----
 src/main/java/net/minecraft/server/PlayerConnection.java     |  2 +-
 .../java/net/minecraft/server/PlayerInteractManager.java     |  2 +-
 src/main/java/org/bukkit/craftbukkit/CraftServer.java        |  3 ++-
 8 files changed, 20 insertions(+), 19 deletions(-)

diff --git a/src/main/java/net/minecraft/server/EntityBoat.java b/src/main/java/net/minecraft/server/EntityBoat.java
index 8fce559..c0695ee 100644
--- a/src/main/java/net/minecraft/server/EntityBoat.java
+++ b/src/main/java/net/minecraft/server/EntityBoat.java
@@ -291,7 +291,7 @@ public class EntityBoat extends Entity {
             }
 
             this.move(this.motX, this.motY, this.motZ);
-            if (this.positionChanged && d3 > 0.2D) {
+            if (!this.dead && this.positionChanged && d3 > 0.2D) { // EMC - Fix wood generator exploit
                 if (!this.world.isStatic) {
                     // CraftBukkit start
                     Vehicle vehicle = (Vehicle) this.getBukkitEntity();
diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 028ba0c..2c59149 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -489,16 +489,9 @@ public abstract class EntityHuman extends EntityLiving implements ICommandListen
     public float a(Block block) {
         float f = this.inventory.a(block);
         int i = EnchantmentManager.getDigSpeedEnchantmentLevel(this);
-        ItemStack itemstack = this.inventory.getItemInHand();
 
-        if (i > 0 && itemstack != null) {
-            float f1 = (float) (i * i + 1);
-
-            if (!itemstack.b(block) && f <= 1.0F) {
-                f += f1 * 0.08F;
-            } else {
-                f += f1;
-            }
+        if (i > 0 && this.inventory.b(block)) {
+            f += (float) (i * i + 1);
         }
 
         if (this.hasEffect(MobEffectList.FASTER_DIG)) {
diff --git a/src/main/java/net/minecraft/server/EntityOcelot.java b/src/main/java/net/minecraft/server/EntityOcelot.java
index 68b08ae..e04c69e 100644
--- a/src/main/java/net/minecraft/server/EntityOcelot.java
+++ b/src/main/java/net/minecraft/server/EntityOcelot.java
@@ -213,7 +213,7 @@ public class EntityOcelot extends EntityTameableAnimal {
     }
 
     public void bG() {
-        if (this.world.random.nextInt(7) == 0) {
+        if (false && this.world.random.nextInt(7) == 0) { // EMC - One Ocelot only...
             for (int i = 0; i < 2; ++i) {
                 EntityOcelot entityocelot = new EntityOcelot(this.world);
 
diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 75c146d..01c8c40 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -303,6 +303,11 @@ public class EntityTrackerEntry {
 
                     this.trackedPlayers.add(entityplayer);
                     Packet packet = this.b();
+                    // EMC start - colored name tags
+                    if (packet != null && packet instanceof Packet20NamedEntitySpawn && this.tracker instanceof EntityPlayer) {
+                        ((Packet20NamedEntitySpawn) packet).b = ((EntityPlayer) this.tracker).listName;
+                    }
+                    // EMC end
 
                     entityplayer.playerConnection.sendPacket(packet);
                     if (!this.tracker.getDataWatcher().d()) {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 79b4369..68f0693 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -352,11 +352,6 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
     public void stop() throws ExceptionWorldConflict { // CraftBukkit - added throws
         if (!this.O) {
             log.info("Stopping server");
-            // CraftBukkit start
-            if (this.server != null) {
-                this.server.disablePlugins();
-            }
-            // CraftBukkit end
 
             if (this.ae() != null) {
                 this.ae().a();
@@ -378,6 +373,13 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
                 worldserver.saveLevel();
             }
             // CraftBukkit end */
+
+            // CraftBukkit start
+            if (this.server != null) {
+                this.server.disablePlugins();
+            }
+            // CraftBukkit end
+
             if (this.n != null && this.n.d()) {
                 this.n.e();
             }
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index b960a87..ef21f89 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -690,7 +690,7 @@ public class PlayerConnection extends Connection {
         }
 
         itemstack = this.player.inventory.getItemInHand();
-        if (itemstack != null && itemstack.count == 0) {
+        if (itemstack != null && itemstack.count <= 0) { // CraftBukkit - Handle infinite use item bug.
             this.player.inventory.items[this.player.inventory.itemInHandIndex] = null;
             itemstack = null;
         }
diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index 55f9ffa..8dc94d4 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -360,7 +360,7 @@ public class PlayerInteractManager {
                 }
             }
 
-            if (itemstack1.count == 0) {
+            if (itemstack1.count <= 0) { // CraftBukkit - Handle infinite use item bug.
                 entityhuman.inventory.items[entityhuman.inventory.itemInHandIndex] = null;
             }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1a3cc03..5c954f7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -277,6 +277,7 @@ public final class CraftServer implements Server {
         if (type == PluginLoadOrder.STARTUP) {
             helpMap.clear();
             helpMap.initializeGeneralTopics();
+            loadCustomPermissions(); // EMC - Plugins should have access to this
         }
 
         Plugin[] plugins = pluginManager.getPlugins();
@@ -289,7 +290,7 @@ public final class CraftServer implements Server {
 
         if (type == PluginLoadOrder.POSTWORLD) {
             commandMap.registerServerAliases();
-            loadCustomPermissions();
+            // loadCustomPermissions(); // EMC - Moving up
             DefaultPermissions.registerCorePermissions();
             helpMap.initializeCommands();
         }
-- 
1.8.1.1

